Source/WebCore/rendering/StyledMarkedText.cpp

```diff
@@ -120,11 +123,82 @@ StyledMarkedText::Style StyledMarkedText::computeStyleForUnmarkedMarkedText(cons
     return style;
 }
 
+static Vector<StyledMarkedText> coalesceAdjacentWithSameRanges(const Vector<StyledMarkedText>& styledTexts)
+{
+    Vector<StyledMarkedText> frontmostMarkedTexts;
+    frontmostMarkedTexts.append(styledTexts[0]);
+    for (auto it = styledTexts.begin() + 1, end = styledTexts.end(); it != end; ++it) {
+        StyledMarkedText& previousStyledMarkedText = frontmostMarkedTexts.last();
+        // StyledMarkedTexts completely cover each other.
+        if (previousStyledMarkedText.startOffset == it->startOffset && previousStyledMarkedText.endOffset == it->endOffset) {
+            // If either background for two different custom highlight StyledMarkedTexts are not opaque, blend colors together.
+            if (previousStyledMarkedText.highlightName != it->highlightName
+                && (!previousStyledMarkedText.style.backgroundColor.isOpaque()
+                    || !it->style.backgroundColor.isOpaque()
+                    || (it->highlightName.isNull() && it->style.backgroundColor.isVisible())))
+                        previousStyledMarkedText.style.backgroundColor = blendSourceOver(previousStyledMarkedText.style.backgroundColor, it->style.backgroundColor);
+            // Else if same priority, take background of the later StyledMarkedText.
+            else if (previousStyledMarkedText.priority == it->priority)
+                previousStyledMarkedText.style.backgroundColor = it->style.backgroundColor;
+            // Take text color of the StyledMarkedText that has set it, maintaining insertion order.
+            if (it->style.textStyles.hasExplicitFillColor)
+                previousStyledMarkedText.style.textStyles.fillColor = it->style.textStyles.fillColor;
+            // Take the highlightName of the latest StyledMarkedText.
+            if (!it->highlightName.isNull())
+                previousStyledMarkedText.highlightName = it->highlightName;
+            // Take the priority of the higher priority StyledMarkedText.
+            if (previousStyledMarkedText.priority < it->priority) {
+                previousStyledMarkedText.priority = it->priority;
+                // If higher priority and opaque, just override background color.
+                if (it->style.backgroundColor.isOpaque())
+                    previousStyledMarkedText.style.backgroundColor = it->style.backgroundColor;
+
+            }
+            continue;
+        }
+        frontmostMarkedTexts.append(*it);
+    }
+    return frontmostMarkedTexts;
+}
+
+static void orderHighlights(const ListHashSet<AtomString>& markedTextsNames, Vector<MarkedText>& markedTexts)
+{
+    if (markedTexts.isEmpty())
+        return;
+
+    HashMap<AtomString, int> markedTextsNamesPriority;
+    unsigned index = 0;
+    for (auto& highlightName : markedTextsNames) {
+        markedTextsNamesPriority.add(highlightName, index);
+        index++;
+    }
+
+    index = 0;
+    while (index < markedTexts.size() - 1) {
```

Comment:
can this just use `std::sort` instead? (not sure if it can or not)

Reply:
I only want to sort given they are highlights, adjacent and cover same ranges. There are non-highlights interspersed in the vector. Is sort able to ignore those? If so, it would be nice to change it to a one-liner.